systems/trajectory_optimization/test/cost_constraint_approximation_test.cc

```diff
@@ -0,0 +1,114 @@
+#include <Eigen/Dense>
+#include <memory>
+#include <utility>
+#include <gtest/gtest.h>
+
+#include "drake/solvers/mathematical_program.h"
+#include "drake/solvers/snopt_solver.h"
+#include "drake/solvers/solve.h"
+#include "systems/trajectory_optimization/dircon_util.h"
+
+namespace dairlib {
+namespace systems {
+namespace {
+
+using drake::solvers::MathematicalProgram;
+using std::cout;
+using std::endl;
+using Eigen::Vector3d;
+using Eigen::VectorXd;
+using Eigen::MatrixXd;
+
+class CostConstraintApproximationTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    // An example
+    H_o_ <<  1.36075, 0.354964,
+         0.354964,  1.19376;
+    b_o_ = VectorXd::Ones(2);
+    c_o_ << 0;
+    A_o_ <<  0.411647, -0.164777,
+         -0.302449, 0.26823;
+    lb_o_ = 0.5 * VectorXd::Ones(2);
+    ub_o_ = VectorXd::Ones(2);
+
+    // solve optimization problem
+    MathematicalProgram quadprog;
+    auto w = quadprog.NewContinuousVariables(2, "w");
+    // intentionally split the following constraint into two constraint bindings
+    quadprog.AddLinearConstraint(A_o_.row(0),
+                                 lb_o_(0),
+                                 ub_o_(0),
+                                 w);
+    quadprog.AddLinearConstraint(A_o_.row(1),
+                                 lb_o_(1),
+                                 ub_o_(1),
+                                 w);
+    // Adds a cost term of the form 0.5*x'*H_o*x + b_o'x + c
+    // intentionally split the following cost into two cost bindings
+    quadprog.AddQuadraticCost(H_o_ * 1 / 3, b_o_ * 1 / 3, w);
+    quadprog.AddQuadraticCost(H_o_ * 2 / 3, b_o_ * 2 / 3, w);
+    const auto result = Solve(quadprog);
+    w_sol_ = result.GetSolution(quadprog.decision_variables());
+
+    // Approximation
+    dairlib::systems::trajectory_optimization::linearizeConstraints(
+        &quadprog, w_sol_, y_a_, A_a_, lb_a_, ub_a_);
+    double c_doub = dairlib::systems::trajectory_optimization::secondOrderCost(
+                      &quadprog, w_sol_, H_a_, b_a_);
+    c_a_ << c_doub;
+  }
+
+  MatrixXd H_o_ = MatrixXd::Zero(2, 2);
```

Comment:
Are these `Zero` intializers necessary? Everything should be set before being read.

Reply:
I think they are necessary. It gave segmentation fault when I didn't add them, so it seems that we need to pre-specify how large the size of the matrix/vector is.

Reply:
But this dosen't matter anymore, cause we are moving them out from setup()